lib/chef/util/path_helper.rb

```diff
@@ -142,6 +142,64 @@ def self.escape_glob(*parts)
+      # If a set of path elements may be provided.  They are appended as-is to the home path if the
+      # homepath exists.  If the home directory doesn't exist or an error is otherwise encountered,
+      # nil is returned.  If an optional block is provided, the joined path is passed to that block
+      # if the home path is valid and the result of the block is returned.
+      def self.home(*args)
+        @@home_dir ||= self.all_homes { |p| break p }
+        if @@home_dir
+          path = File.join(@@home_dir, *args)
+          block_given? ? (yield path) : path
+        end
+      end
+
+      # See self.home.  This method performs a similar operation except that it yields all the different
+      # possible values of 'HOME' that one could have on this platform.  Hence, on windows, if
+      # HOMEDRIVE\HOMEPATH and USERPROFILE are different, the provided block will be called twice.
+      # This method goes out and checks the existence of each location at the time of the call.
+      #
+      # The return is a list of all the returned values from each block invocation or a list of paths
+      # if no block is provided.
+      def self.all_homes(*args)
+        paths = []
+        if Chef::Platform.windows?
+          # By default, Ruby uses the the following environment variables to determine Dir.home:
+          # HOME
+          # HOMEDRIVE HOMEPATH
+          # USERPROFILE
+          # Ruby only checks to see if the variable is specified - not if the directory actually exists.
+          # On Windows, HOMEDRIVE HOMEPATH can point to a different location (such as an unavailable network mounted drive)
+          # while USERPROFILE points to the location where the user application settings and profile are stored.  HOME
+          # is not defined as an environment variable (usually).  If the home path actually uses UNC, then the prefix is
+          # HOMESHARE instead of HOMEDRIVE.
+          #
+          # We instead walk down the following and if that path actually exists, we pick and use that.  Once a home-path
+          # has been picked, that value is memoized so that subsequent calls to home_dir don't bounce around.
+          # HOME
+          # HOMEDRIVE HOMEPATH
+          # HOMESHARE HOMEPATH
+          # USERPROFILE
+
+          paths << ENV['HOME']
+          paths << ENV['HOMEDRIVE'] + ENV['HOMEPATH'] if ENV['HOMEDRIVE'] && ENV['HOMEPATH']
+          paths << ENV['HOMESHARE'] + ENV['HOMEPATH'] if ENV['HOMESHARE'] && ENV['HOMEPATH']
+          paths << ENV['USERPROFILE']
+        end
+        paths << Dir.home
+
+        (paths.reduce([]) do |acc, home_path|
+          home_path = home_path.gsub(path_separator, ::File::SEPARATOR) if home_path
+          acc << home_path if home_path && Dir.exists?(home_path) && !acc.include?(home_path)
+          acc
+        end).map do |home_path|
+          joined_path = File.join(home_path, *args)
+          block_given? ? (yield joined_path) : joined_path
```

Comment:
I was implementing allowing for an optional &block to be provided.  I don't know how idiomatic it is in ruby to do something like PathHelper.home {|p| puts(p) } - i.e. the provided block only gets called if a resource it needs is actually valid/alive.  I was modeling it off of the "with" clause in Python or the "using" clause in C#.